src/array_api_extra/_funcs.py

```diff
@@ -546,3 +552,275 @@ def sinc(x: Array, /, *, xp: ModuleType | None = None) -> Array:
+    on read-only versus writeable arrays::
+
+        x = xp.asarray([0, 0, 0])
+        y = at(x, 0).set(2, copy=None)
+        z = at(x, 1).set(3, copy=None)
+
+    In the above example, ``x == [0, 0, 0]``, ``y == [2, 0, 0]`` and z == ``[0, 3, 0]``
+    when x is read-only, whereas ``x == y == z == [2, 3, 0]`` when x is writeable!
+
+    Warning
+    -------
+    The behaviour of update methods when the index is an array of integers which
+    contains multiple occurrences of the same index is undefined;
+    e.g. ``at(x, [0, 0]).set(2)``
+
+    Note
+    ----
+    `sparse <https://sparse.pydata.org/>`_ is not supported by update methods yet.
+
+    See Also
+    --------
+    `jax.numpy.ndarray.at <https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ndarray.at.html>`_
+    """
+
+    x: Array
+    idx: Any
+    __slots__ = ("x", "idx")
+
+    def __init__(self, x: Array, idx: Any = _undef, /):
+        self.x = x
+        self.idx = idx
+
+    def __getitem__(self, idx: Any) -> Any:
+        """Allow for the alternate syntax ``at(x)[start:stop:step]``,
+        which looks prettier than ``at(x, slice(start, stop, step))``
+        and feels more intuitive coming from the JAX documentation.
+        """
+        if self.idx is not _undef:
+            raise ValueError("Index has already been set")
+        self.idx = idx
+        return self
+
+    def _common(
+        self,
+        at_op: str,
+        y: Array = _undef,
+        /,
+        copy: bool | None | Literal["_force_false"] = True,
+        xp: ModuleType | None = None,
+        _is_update: bool = True,
+        **kwargs: Any,
+    ) -> tuple[Any, None] | tuple[None, Array]:
+        """Perform common prepocessing.
+
+        Returns
+        -------
+        If the operation can be resolved by at[], (return value, None)
+        Otherwise, (None, preprocessed x)
+        """
+        if self.idx is _undef:
+            raise TypeError(
+                "Index has not been set.\n"
+                "Usage: either\n"
+                "    at(x, idx).set(value)\n"
+                "or\n"
+                "    at(x)[idx].set(value)\n"
+                "(same for all other methods)."
+            )
+
+        x = self.x
+
+        if copy is True:
+            writeable = None
+        elif copy is False:
+            writeable = is_writeable_array(x)
+            if not writeable:
+                raise ValueError("Cannot modify parameter in place")
+        elif copy is None:
+            writeable = is_writeable_array(x)
+            copy = _is_update and not writeable
+        elif copy == "_force_false":  # type: ignore[redundant-expr]
+            # __getitem__ with fancy index on a numpy array
+            writeable = True
+            copy = False
+        else:
+            raise ValueError(f"Invalid value for copy: {copy!r}")
+
+        if copy:
+            try:
+                at_ = x.at
+            except AttributeError:
```

Comment:
The alternative to this duck-type pattern is to import `is_jax_array` from array-api-compat. @lucascolley please advise on what you'd prefer.